File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        chars=set(s)
        
        answer=0
        done=set()
        for i in chars:
            new=set()
            Found=False
            for j in range(len(s)):
                if s[j]==i and Found:
                    
                    if len(new)!=0:
                        done.update(new)
                        new=set()
                        new.add(i)
                    else:
                        new.add(i)
                    
                elif s[j]==i:
                    Found=True
                elif Found:
                    if s[j] not in done:
                        new.add(s[j])
            answer+=len(done)
            done=set()
        return answer


                    






        

File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.java
class Solution {
    public int countPalindromicSubsequence(String s) {

        // Get distinct characters in the string
        Set<Character> chars = new HashSet<>();
        for (char c : s.toCharArray()) {
            chars.add(c);
        }

        int answer = 0;

        // For each distinct character
        for (char c : chars) {
            Set<Character> done = new HashSet<>();
            Set<Character> temp = new HashSet<>();
            boolean found = false;

            for (int j = 0; j < s.length(); j++) {
                char ch = s.charAt(j);

                // If same char found again after first occurrence
                if (ch == c && found) {
                    if (!temp.isEmpty()) {
                        done.addAll(temp);
                        temp.clear();
                        temp.add(c);
                    } else {
                        temp.add(c);
                    }
                }
                // First time encountering c
                else if (ch == c) {
                    found = true;
                }
                // Characters between the two occurrences of c
                else if (found) {
                    if (!done.contains(ch)) {
                        temp.add(ch);
                    }
                }
            }

            answer += done.size();
        }

        return answer;
    }
}
